subsys/bluetooth/mesh/adv.c

```diff
@@ -93,35 +95,126 @@ static struct net_buf *bt_mesh_adv_create_from_pool(struct net_buf_pool *buf_poo
 	adv->type         = type;
 	adv->tag          = tag;
 	adv->xmit         = xmit;
+	adv->prio         = prio;
 
 	return buf;
 }
 
-struct net_buf *bt_mesh_adv_create(enum bt_mesh_adv_type type,
-				   enum bt_mesh_adv_tag tag,
-				   uint8_t xmit, k_timeout_t timeout)
+struct net_buf *bt_mesh_adv_relay_create(uint8_t prio, uint8_t xmit)
 {
 #if defined(CONFIG_BT_MESH_RELAY)
-	if (tag & BT_MESH_RELAY_ADV) {
-		return bt_mesh_adv_create_from_pool(&relay_buf_pool,
-						    adv_relay_pool, type,
-						    tag, xmit, timeout);
+	sys_sfnode_t *prev, *curr, *lowest_prev = NULL, *lowest = NULL;
+	uint8_t prio_cur = prio;
+	struct net_buf *buf;
+
+	buf = bt_mesh_adv_create_from_pool(&relay_buf_pool, adv_relay_pool,
+					   BT_MESH_ADV_DATA, BT_MESH_RELAY_ADV,
+					   xmit, prio, K_NO_WAIT);
+	if (buf) {
+		return buf;
 	}
-#endif
 
-	return bt_mesh_adv_create_from_pool(&adv_buf_pool, adv_local_pool, type,
-					    tag, xmit, timeout);
+	if (!IS_ENABLED(CONFIG_BT_MESH_RELAY_PRIORITY) || !prio_cur) {
+		return NULL;
+	}
+
+	/* Try to release the buffer from the low priority queue to the high priority message.
+	 * For the same low priority message, the one that takes the longest buffer time will
+	 * be released first.
+	 */
+	k_mutex_lock(&bt_mesh_relay_mutex, K_FOREVER);
```

Comment:
I don't like to have mutex with sleep at the bottom layer. The mesh code is not thread-safe. If reschedule happens and we enter mesh code from another thread (this is currently possible because ble host calls mesh from a separate thread: https://github.com/zephyrproject-rtos/zephyr/blob/87a751c266da02b53193852e336c3227bf7a3672/subsys/bluetooth/host/Kconfig#L86-L89), this can mess up states.

Reply:
I also thought about this situation, but ensuring thread safety is a problem. In the design of zephyr bluetooth, it is a conventional limitation to use Co-Operative.

``` C
	k_thread_create(&tx_thread_data, tx_thread_stack,
			K_KERNEL_STACK_SIZEOF(tx_thread_stack),
			hci_tx_thread, NULL, NULL, NULL,
			K_PRIO_COOP(CONFIG_BT_HCI_TX_PRIO),
			0, K_NO_WAIT);
	k_thread_name_set(&tx_thread_data, "BT TX");

#if defined(CONFIG_BT_RECV_WORKQ_BT)
	/* RX thread */
	k_work_queue_init(&bt_workq);
	k_work_queue_start(&bt_workq, rx_thread_stack,
			   CONFIG_BT_RX_STACK_SIZE,
			   K_PRIO_COOP(CONFIG_BT_RX_PRIO), NULL);
	k_thread_name_set(&bt_workq.thread, "BT RX");
#endif
```

Reply:
So there is no need for mutex before.